File: RetiDiCalcolatori/Estensioni/es00_ext/Produttore.java
                                                                                                                                                                                                                                                                                                                                                                                                                             import java.io.BufferedReader;
import java.io.FileWriter;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.PrintWriter;

public class Produttore {
    /*specifiche:
    Il produttore non chiede all’utente quante righe scrivere, ma è un
    programma sequenziale che legge fino a quando l’utente immette
    EOF (end of file) per terminare l’inserimento (è un filtro)
    Ciascuna riga letta comincia con un prefisso composto da un intero
    seguito dal separatore ‘:’; tale intero indica all’interno di quale file
    andrà scritto (in append) il contenuto della riga letta ad esempio:
    >1:Questa riga andra’ in fileName1.txt
    // in fileName1.txt, che assumiamo essere il primo argomento
    >3:Questa riga andra’ in fileName3.txt
    // in fileName3.txt, che assumiamo essere il terzo argomento
    >1:Questa riga andra’ in fileName1.txt
    …
    Una volta terminato il ciclo di letture il produttore chiude tutti i file
    aperti e libera tutte le risorse occupate
     */
  
    public static void main(String[] args) {
        if (args.length < 1) {
            System.err.println("Errore: numero di argomenti insufficienti");
            System.exit(1);
        }
        int numFiles, i;
        numFiles = args.length;
        String filename = null;
        PrintWriter[] writers = new PrintWriter[numFiles];
        //dichiaro le variabili prima di entrare dentro il ciclo for
        
        // Apri tutti i file in modalità append
        try {
            for (i = 0; i < numFiles; i++) {
                filename = args[i];
                
                // Controllo se il file è un .txt
                if (!filename.endsWith(".txt")) {
                    System.err.println("Errore: " + filename + " non è un file .txt valido.");
                    System.exit(1);
                }
                // Apri il file in modalità append
                writers[i] = new PrintWriter(new FileWriter(filename, true));
            }

            BufferedReader stdIn = null;
            int indice, sepIndex;
            String testo, indiceStr, riga;
            indice =-1;
            sepIndex = -1;
            testo = null;
            indiceStr = null;
            riga = null;
            //dichiaro le variabili prima di entrare nel ciclo while

            System.out.println("Scrivere le righe da inserire (CTRL+D o CTRL+Z per terminare):");
		    stdIn = new BufferedReader(new InputStreamReader(System.in));
     
            while ((riga = stdIn.readLine())!=null) {
                
                // Trova il separatore ':'
                sepIndex = riga.indexOf(':');
                if (sepIndex == -1) {
                    System.err.println("Formato errato: manca il separatore ':' in: " + riga);
                    for (i = 0;i<writers.length;i++) {
                        if (writers[i] != null) {
                            writers[i].close();
                        }
                    }
                    System.exit(1);
                }
                
                // Estrai indice numerico
                indiceStr = riga.substring(0, sepIndex);
                
                // Verifica che sia numerico
                if (!indiceStr.matches("\\d+")) {
                    System.err.println("Formato errato: prima del separatore deve esserci un indice numerico in: " + riga);
                    for (i = 0;i<writers.length;i++) {
                        if (writers[i] != null) {
                            writers[i].close();
                        }
                    }
                    System.exit(1);
                }
                try{
                indice = Integer.parseInt(indiceStr);
                }catch(NumberFormatException e){
                    System.err.println("Formato errato: prima del separatore deve esserci un indice numerico in: " + riga);
                    for (i = 0;i<writers.length;i++) {
                        if (writers[i] != null) {
                            writers[i].close();
                        }
                    }
                    System.exit(1);
                }
                testo = riga.substring(sepIndex + 1);
                
                if (indice < 1 || indice > numFiles) {
                    System.err.println("Indice file non valido: " + indice);
                    for (i = 0;i<writers.length;i++) {
                        if (writers[i] != null) {
                            writers[i].close();
                        }
                    }
                    System.exit(1);
                }
                
                // Scrivi nel file appropriato
                writers[indice-1].append(testo+System.lineSeparator());
                writers[indice-1].flush();
            }
                        
        } catch (IOException e) {
            System.err.println("Errore I/O: " + e.getMessage());
            for (i = 0;i<writers.length;i++) {
                if (writers[i] != null) {
                    writers[i].close();
                }
            }
            System.exit(1);
        } catch (NumberFormatException e) {
            System.err.println("Errore formato numero: " + e.getMessage());
            for (i = 0;i<writers.length;i++) {
                if (writers[i] != null) {
                    writers[i].close();
                }
            }
            System.exit(1);
        }
        
        // Chiudi tutti i file
        for (i = 0;i<writers.length;i++) {
            if (writers[i] != null) {
                writers[i].close();
            }
        }
        System.out.println("Tutti i file chiusi correttamente.");
    }
    

    
}

File: RetiDiCalcolatori/Estensioni/es00_ext/produttore.c
#include <fcntl.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#define MAX_STRING_LENGTH 256
/*specifiche:
    Il produttore non chiede all’utente quante righe scrivere, ma è un
    programma sequenziale che legge fino a quando l’utente immette
    EOF (end of file) per terminare l’inserimento (è un filtro)
    Ciascuna riga letta comincia con un prefisso composto da un intero
    seguito dal separatore ‘:’; tale intero indica all’interno di quale file
    andrà scritto (in append) il contenuto della riga letta ad esempio:
    >1:Questa riga andra’ in fileName1.txt
    // in fileName1.txt, che assumiamo essere il primo argomento
    >3:Questa riga andra’ in fileName3.txt
    // in fileName3.txt, che assumiamo essere il terzo argomento
    >1:Questa riga andra’ in fileName1.txt
    …
    Una volta terminato il ciclo di letture il produttore chiude tutti i file
    aperti e libera tutte le risorse occupate
*/
int main(int argc, char *argv[]) {
    if (argc < 2) {
        printf("Errore: numeri di argomenti insufficienti per l'esecuzione di %s", argv[0]);
        exit(EXIT_FAILURE);
    }
    int num_files, fd[argc-1], i, j, indice, len, written;
    char riga[MAX_STRING_LENGTH], *sep, *testo, filename[MAX_STRING_LENGTH];

    num_files= argc-1;

    // Apri tutti i file in modalità append
    for (i=0; i<num_files; i++) {
        strcpy(filename,argv[i+1]);
        len = strlen(filename);
        //controllo del file se sia un .txt
        if (len<5 || strcmp(filename+len-4, ".txt") != 0) {
            printf("Errore: %s non è un file .txt valido.\n", filename);
            exit(EXIT_FAILURE);
        }
        //controllo permessi di scrittura
        if (access(filename,W_OK)!=0) {
            perror("File permesso negato");
            exit(EXIT_FAILURE);
        }
        fd[i]= open(filename, O_WRONLY | O_CREAT | O_APPEND, 0640);
        if (fd[i]<0) {
            perror("Errore apertura file");
            // Chiudi eventuali file già aperti
            for (j=0;j<i;j++) close(fd[j]);
            exit(EXIT_FAILURE);
        }
    }
    printf("scrivere le righe da inserire\n");
    while (gets(riga)) {
        // Rimuovi eventuale newline finale
        riga[strlen(riga)+1] = '\0';
        riga[strlen(riga)] = '\n';

        // Trova il separatore ':'
        sep = strchr(riga, ':');
        if (!sep) {
            printf("Formato errato: manca il separatore ':'  %s\n", riga);
            exit(EXIT_FAILURE);
        }

        // Estrai indice numerico
        *sep = '\0'; // separa la stringa in due parti
        for(j=0;riga[j]!='\0';j++){
            if(riga[j] < '0' || riga[j] > '9')
                {
                    printf("Formato errato: prima del separatore deve esserci un indice numerico %s\n", riga);
                    exit(EXIT_FAILURE);
                }
        }
        indice = atoi(riga); // parte prima dei due punti
        testo = sep + 1; // assegno al puntatore il puntatore dopo i due punti

        if (indice<1 || indice>num_files) {
            printf("Indice file non valido: %d\n", indice);
            exit(EXIT_FAILURE);

        }

            // Aggiungi newline se non presente
            len = strlen(testo);
            if (len == 0) {
                testo[len] = '\n';
                testo[len + 1] = '\0';
            }

        written = write(fd[indice-1], testo, sizeof(char)*(len));
        if (written < 0) {
            perror("Errore scrittura file");
            exit(EXIT_FAILURE);
        }
    }

    // Chiudi tutti i file
    for (i=0; i<num_files; i++) {
        close(fd[i]);
    }

    printf("Tutti i file chiusi correttamente.\n");
    return EXIT_SUCCESS;
}
